t/semaphore.t

```diff
@@ -0,0 +1,1068 @@
+                if not g.test then
+                    local sem, err = semaphore.new(0)
+                    if not sem then
+                        ngx.log(ngx.ERR, err)
+                    end
+                    g.test = sem
+                end
+                local sem = g.test
+                local ok, err = sem:post()
+                if not ok then
+                    ngx.log(ngx.ERR, err)
+                end
+            end
+            ngx.timer.at(0, func,g)
+            ngx.sleep(2)
+            ngx.print("post")
+            ngx.exit(200)
+    ';
+    }
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+200
+wait
+200
+post
+--- log_level: debug
+--- no_error_log
+[error]
+
+
+
+=== TEST 18: a light thread that to be killed is waitting a semaphore
+--- http_config eval: $::HttpConfig
+--- config
+    location /test {
+       content_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local sem = semaphore.new(0)
+            if not sem then
+                error("create failed")
+            end
+            local function func(sem)
+                sem:wait(10)
+            end
+            local co = ngx.thread.spawn(func,sem)
+            ngx.thread.kill(co)
+            ngx.say("test")
+       ';
+    }
+--- log_level: debug
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+test
+--- error_log
+ngx_http_lua_semaphore_cleanup
+
+
+
+=== TEST 19: a light thread that is going to exit is waitting a semaphore
+--- http_config eval: $::HttpConfig
+--- config
+    location /test {
+       content_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local sem = semaphore.new(0)
+            if not sem then
+                error("create failed")
+            end
+            local function func(sem)
+                sem:wait(10)
+            end
+            local co = ngx.thread.spawn(func,sem)
+            ngx.say("test")
+            ngx.exit(200)
+       ';
+    }
+--- log_level: debug
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+test
+--- error_log
+ngx_http_lua_semaphore_cleanup
+
+
+
+=== TEST 20: test thread wait or post a sem in one request
+--- http_config eval: $::HttpConfig
+--- config
+    location = /test {
+        access_log off;
+        content_by_lua '
+            local semaphore = require "ngx.semaphore"
+
+            local g = getmetatable(_G).__index
+            local sem, err = semaphore.new(0)
+            if not sem then
+                ngx.say(err)
+                ngx.exit(500)
+            end
+            g.test = sem
+
+            local sem2, err = semaphore.new(0)
+            if not sem2 then
+                ngx.say(err)
+                ngx.exit(500)
+            end
+            g.test_next = sem2
+
+            g.log = {}
+            g.log_i = 0
+            g.wait_id = 0
+            g.wakeup_id = 0
+            --0:wait 1:post 2:post_all
+            g.test_op = {0,1,1,0}
```

Comment:
Let's avoid such magical numbers. In Lua, strings are always intern'd and thus as cheap as "symbols" in other programming languages. So just use strings here to avoid confusions. Tests should be simple to stay away from bugs in themselves.